packages/eslint-plugin/lib/rules/await-promise.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * @fileoverview Disallows awaiting a value that is not a Promise
+ * @author Josh Goldberg
+ */
+'use strict';
+const util = require('../util');
+const types = require('../utils/types');
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const defaultOptions = [
+  {
+    allowedPromiseNames: []
+  }
+];
+
+/**
+ * @type {import("eslint").Rule.RuleModule}
+ */
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Disallows awaiting a value that is not a Promise',
+      category: 'Functionality',
+      recommended: 'error',
+      extraDescription: [util.tslintRule('await-promise')],
+      url: util.metaDocsUrl('await-promise')
+    },
+    fixable: null,
+    messages: {
+      await: 'Invalid `await` of a non-Promise value.',
+      forOf: 'Invalid `for-await-of` of a non-AsyncIterable value.'
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          allowedPromiseNames: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          }
+        },
+        additionalProperties: false
+      }
+    ],
+    type: 'problem'
+  },
+
+  create(context) {
+    const options = util.applyDefault(defaultOptions, context.options)[0];
+
+    const allowedAsyncIterableNames = new Set([
+      'AsyncIterable',
+      'AsyncIterableIterator'
```

Comment:
I'd like to see tests for objects that implements `[Symbol.asyncIterator]` property. For example, we can iterate `stream.Readable` with `for-await-of` syntax.

Reply:
@mysticatea I would also like to see those! But: my understanding is that that would be quite tricky, as we'd have to check every base type of the object's type for implementing the `[Symbol.asyncIterator]` property correctly, and making sure they do it correctly accounting for call signatures (unless I'm wrong - am I?). Is there a good way I'm missing?

TSLint right now does approximately what this implementation does - allow a whitelist of names. If the `stream.Readable` class is declared as implementing any of those names it'll be allowed.

In lieu of doing this correctly, or waiting for https://github.com/Microsoft/TypeScript/issues/9879, would it be all right if this just accepted more `allowedAsyncIterableNames` as an option? _(or some better option name than that?)_

Reply:
Because I'm not familiar with TypeScript compiler API, I'm not sure if the way exists or not. I will investigate it.

But I think that the current this rule is too strict and lacking the way which to configure to use streams with `for-await-of` syntax.

Reply:
@mysticatea coming back to this, I think it's actually misplaced to have async iterable checking in this rule. `await-promise` as a rule name doesn't imply that it should check async iterables.

I've removed the async iterable checking in the latest commit. Could we merge this rule in now for feature compatibility with TSLint, then work on them in a separate issue? I'd be happy to send a PR with a separate rule for them.

Reply:
Thank you.

In that case, I don't oppose to merge. Though, it's better if the rule checks the existence of `then` member rather than type names. If so, we can remove options from this rule.

I'm sorry that I'm delaying to investigate.